### PR TITLE
WRO-14592: [TransferList POC] - Disable items or Allow transfer functionality

### PR DESCRIPTION
**Pull request by adrian-cocoara-lgp**
Enact-DCO-1.0-Signed-off-by: Adrian Cocoara adrian.cocoara@lgepartner.com

### Checklist

* [x] I have read and understand the [contribution guide](http://enactjs.com/docs/developer-guide/contributing/)
* [ ] A [CHANGELOG entry](http://enactjs.com/docs/developer-guide/contributing/changelogs/) is included
* [ ] At least one test case is included for this feature or bug fix
* [x] Documentation was added or is not needed
* [ ] This is an API breaking change

### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)
Added disabled prop and min-max capacity for transfer lists

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)

### Additional Considerations
[//]: # (How should the change be tested?)
[//]: # (Are there any outstanding questions?)
[//]: # (Were any side-effects caused by the change?)

### Links
[//]: # (Related issues, references)
WRO-14592

### CODE REVIEW

File: TransferList/TransferList.js [hunk 1]
```diff
@@ -71,6 +80,22 @@ const TransferListBase = kind({
 		 */
 		firstList: PropTypes.array,
 
+		/**
+		 * Sets the maximum number of items for the first list.
+		 *
+		 * @type {Number}
+		 * @public
+		 */
+		firstListMaximumCapacity: PropTypes.number,
```
Comment (stanca-pop-lgp) [inline TransferList/TransferList.js:89]: I suggest a minor abbreviation 'firstListMaxCapacity' and 'firstListMinCapacity'. Same for secondList.

File: TransferList/TransferList.js [hunk 2]
```diff
@@ -423,7 +473,7 @@ const TransferListBase = kind({
 			setNewList(list);
 		};
 
-		const rearrangeLists = (sourceList, destinationList, draggedElementIndex, draggedElementList, dragOverElementIndex, setSourceList, setDestinationList) => {
+		const rearrangeLists = useCallback((sourceList, destinationList, draggedElementIndex, draggedElementList, dragOverElementIndex, setSourceList, setDestinationList) => {
```
Comment (adrian-cocoara-lgp) [inline TransferList/TransferList.js:476]: Note: `rearrangeLists` was refactored to `useCallback` as a consequence of refactoring `onDropRightHandler` and `onDropLeftHandler` to use `useCalback`.
image

File: TransferList/TransferList.js [hunk 3]
```diff
@@ -402,7 +451,7 @@ const TransferListBase = kind({
 				setSecondListLocal(tempSecond);
 			}
 			setSelectedItems(tempSelected);
-		}, [firstListLocal, secondListLocal, selectedItems, setFirstList, setSecondList]);
+			}, [firstListLocal, firstListMinimumCapacity, secondListLocal, selectedItems, setFirstList, setSecondList, secondListMaximumCapacity]);
```
Comment (stanca-pop-lgp) [inline TransferList/TransferList.js:454]: Please check indentation.

**Review by stanca-pop-lgp — CHANGES_REQUESTED** [review]

File: TransferList/TransferList.js [hunk 4]
```diff
@@ -587,10 +641,10 @@ const TransferListBase = kind({
 				<Cell className={componentCss.listButtons}>
 					{!moveOnSpotlight ?
 						<>
-							<Button onClick={moveIntoSecondAll} onSpotlightUp={handleSpotlightBounds} size="small">{'>>>'}</Button>
-							<Button disabled={!(selectedItems.find((item) => item.list === "first"))} onClick={moveIntoSecondSelected} size="small">{'>'}</Button>
-							<Button disabled={!(selectedItems.find((item) => item.list === "second"))} onClick={moveIntoFirstSelected} size="small">{'<'}</Button>
-							<Button onClick={moveIntoFirstAll} size="small">{'<<<'}</Button>
+							<Button disabled={disabled || (secondListMaximumCapacity || firstListMinimumCapacity) !== undefined} onClick={moveIntoSecondAll} onSpotlightUp={handleSpotlightBounds} size="small">{'>>>'}</Button>
+							<Button disabled={!(selectedItems.find((item) => item.list === "first")) || disabled} onClick={moveIntoSecondSelected} size="small">{'>'}</Button>
```
Comment (daniel-stoian-lgp) [inline TransferList/TransferList.js:645]:
```suggestion
							<Button disabled={disabled || !!secondListMaximumCapacity || !!firstListMinimumCapacity} onClick={moveIntoSecondAll} onSpotlightUp={handleSpotlightBounds} size="small">{'>>>'}</Button>
```

File: TransferList/TransferList.js [hunk 5]
```diff
@@ -587,10 +641,10 @@ const TransferListBase = kind({
 				<Cell className={componentCss.listButtons}>
 					{!moveOnSpotlight ?
 						<>
-							<Button onClick={moveIntoSecondAll} onSpotlightUp={handleSpotlightBounds} size="small">{'>>>'}</Button>
-							<Button disabled={!(selectedItems.find((item) => item.list === "first"))} onClick={moveIntoSecondSelected} size="small">{'>'}</Button>
-							<Button disabled={!(selectedItems.find((item) => item.list === "second"))} onClick={moveIntoFirstSelected} size="small">{'<'}</Button>
-							<Button onClick={moveIntoFirstAll} size="small">{'<<<'}</Button>
+							<Button disabled={disabled || (secondListMaximumCapacity || firstListMinimumCapacity) !== undefined} onClick={moveIntoSecondAll} onSpotlightUp={handleSpotlightBounds} size="small">{'>>>'}</Button>
```
Comment (daniel-stoian-lgp) [inline TransferList/TransferList.js:644]:
```suggestion
							<Button disabled={disabled || !!secondListMaximumCapacity || !!firstListMinimumCapacity} onClick={moveIntoSecondAll} onSpotlightUp={handleSpotlightBounds} size="small">{'>>>'}</Button>
```

File: TransferList/TransferList.js [hunk 6]
```diff
@@ -587,10 +641,10 @@ const TransferListBase = kind({
...
+							<Button disabled={!(selectedItems.find((item) => item.list === "first")) || disabled} onClick={moveIntoSecondSelected} size="small">{'>'}</Button>
+							<Button disabled={!(selectedItems.find((item) => item.list === "second")) || disabled} onClick={moveIntoFirstSelected} size="small">{'<'}</Button>
+							<Button disabled={disabled || (firstListMaximumCapacity || secondListMinimumCapacity) !== undefined} onClick={moveIntoFirstAll} size="small">{'<<<'}</Button>
```
Comment (daniel-stoian-lgp) [inline TransferList/TransferList.js:647]: same logic here. we don't need to use undefined. please remove from eslint-disable rules

File: TransferList/TransferList.js (hunk 1 above)
Comment (adrian-cocoara-lgp) [reply TransferList/TransferList.js:89]: ok

**Review by stanca-pop-lgp — APPROVED** [review]

**Review by daniel-stoian-lgp — APPROVED** [review]

**Pull request merged**